src/AutoRest.CSharp/Output/Models/Types/OutputLibrary.cs

```diff
@@ -22,21 +25,65 @@ internal class OutputLibrary
         private Dictionary<OperationGroup, RestClient>? _restClients;
         private Dictionary<OperationGroup, ResourceOperation>? _resourceOperations;
         private Dictionary<OperationGroup, ResourceContainer>? _resourceContainers;
+        private Dictionary<Schema, TypeProvider>? _resourceModels;
         private Dictionary<Operation, LongRunningOperation>? _operations;
         private Dictionary<Operation, ResponseHeaderGroupType>? _headerModels;
+        private Dictionary<string, List<OperationGroup>> _operationGroups;
+
+        private Dictionary<Schema, TypeProvider> SchemaMap
+        {
+            get
+            {
+                if (_models is null)
+                {
+                    BuildModels();
+                }
+#pragma warning disable CS8603 // Possible null reference return.
+                return _models;
+#pragma warning restore CS8603 // Possible null reference return.
+            }
+        }
+
+        private Dictionary<Schema, TypeProvider> ResourceSchemaMap
+        {
+            get
+            {
+                if (_resourceModels is null)
+                {
+                    BuildModels();
+                }
+#pragma warning disable CS8603 // Possible null reference return.
+                return _resourceModels;
+#pragma warning restore CS8603 // Possible null reference return.
+            }
+        }
 
         public OutputLibrary(CodeModel codeModel, BuildContext context)
         {
             _codeModel = codeModel;
             _context = context;
+            _operationGroups = new Dictionary<string, List<OperationGroup>>();
             if (context.Configuration.AzureArm)
             {
                 DecorateOperationGroup();
+
+                List<OperationGroup>? result;
+                foreach (var operationGroup in _codeModel.OperationGroups)
+                {
+                    if (!_operationGroups.TryGetValue(operationGroup.Resource, out result))
```

Comment:
Move this into the DecorateOperationGroup Method loop and put the body of this for in a method called AddOperationGroupToResourceMap